docs/proposals/read-write-operations-bucket.md

```diff
@@ -0,0 +1,244 @@
+# Read-Write coordination free operational contract for object storage
+
+Status: draft | **in-review** | rejected | accepted | complete
+
+Implementation Owner: [@bwplotka](https://github.com/bwplotka)
+
+Tickets: 
+* https://github.com/improbable-eng/thanos/issues/298 (eventual consistency)
+* https://github.com/improbable-eng/thanos/issues/377 (eventual consistency & partial upload)
+* https://github.com/improbable-eng/thanos/issues/564 (retention vs store gateway)
+* https://github.com/improbable-eng/thanos/issues/271 (adapt for removals faster)
+
+## Summary
+
+Our goals here are:
+A) Define consistent way of having multiple readers and multiple writers (technically multiple appenders and single remover) on shared object storage that can be 
+eventual consistent. 
+B) Allow readers to sync the object storage changes in eventual consistent manner: (e.g to poll for changes 
+every X minutes instead of watch and react to changes immediately)
+C) Allow readers to detect and handle partial uploads consistently without any coordination.
+
+## Motivation 
+
+Thanos performs similar operations as Prometheus do on TSDB blocks with the following differences:
+* It operates on Object Storage API instead of local filesystem.
+* Operations are done from multiple processes that lack coordination.
+
+Moving from lock-based logic to coordination free and from strongly consistent local filesystem to potentially eventually 
+consistent remote simplified "filesystem" in form of Object Storage API, causes additional cases that we need to consider 
+in Thanos system, like:
+* Thanos sidecar or compactor crashes during the process of uploading the block. It uploaded index, 2 chunk files and crashed. How to ensure readers (e.g compactor, store gateway) will handle this gracefully?
+* Thanos compactor uploads compacted block and deletes source blocks. After next sync iteration it does not see a new block (read after write eventual consistency). It sees gap, wrongly plans next compaction and causes non-resolvable overlap.
+* Thanos compactor uploads compacted block and deletes source blocks. Thanos Store Gateway syncs every 3m so it missed that fact. Next query that hits store gateway tries to fetch deleted source blocks and fails.
+
+Currently, we have only basic safeguards against those in form of [`syncDelay`](https://github.com/improbable-eng/thanos/blob/bc088285a1b4bf464fdf2539e4b365b805874eed/pkg/compact/compact.go#L187). 
+That helps with operations between sidecar and compactor but does not fix inter-compactor logic in case of eventual consistent storage or partial upload failures.
+
+We also partially assumed strong consistency for object storage, because [GCS has strong consistency](https://cloud.google.com/storage/docs/consistency) and S3 has as it as well, 
+but with [some caveats](https://codeburst.io/quick-explanation-of-the-s3-consistency-model-6c9f325e3f82) that we met.
```

Comment:
So, does it have strong consistency or not? Or more importantly, do you want to say that effort has been made to such that our use of S3 is within it's strongly consistent guarantees?

Reply:
Yes, indeed. There was an effort to eliminate invokes `Get` or `Exists` before writing object in crucial parts, but we DID not eliminate them everywhere, that's why some `syncDelay` exists for only part of new blocks.